src/data/matrix/kronecker.lean

```diff
@@ -0,0 +1,319 @@
+  [module R α] [module R β] [module R γ]
+  (f : α →ₗ[R] β →ₗ[R] γ) :
+  matrix l m α →ₗ[R] matrix n p β →ₗ[R] matrix (l × n) (m × p) γ :=
+linear_map.mk₂ R
+  (kronecker_map (λ r s, f r s))
+  (kronecker_map_add_left _ $ f.map_add₂)
+  (kronecker_map_smul_left _ $ f.map_smul₂)
+  (kronecker_map_add_right _ $ λ a, (f a).map_add)
+  (kronecker_map_smul_right _ $ λ r a, (f a).map_smul r)
+
+/-- `matrix.kronecker_map_linear` commutes with `⬝` if `f` commutes with `*`.
+
+This is primarily used with `R = ℕ` to prove `matrix.mul_kronecker_mul`. -/
+lemma kronecker_map_linear_mul_mul [comm_semiring R]
+  [non_unital_non_assoc_semiring α] [non_unital_non_assoc_semiring β]
+  [non_unital_non_assoc_semiring γ]
+  [module R α] [module R β] [module R γ]
+  (f : α →ₗ[R] β →ₗ[R] γ) (h_comm : ∀ a b a' b', f (a * b) (a' * b') = f a a' * f b b')
+  (A : matrix l m α) (B : matrix m n α) (A' : matrix l' m' β) (B' : matrix m' n' β) :
+  kronecker_map_linear f (A ⬝ B) (A' ⬝ B') =
+    (kronecker_map_linear f A A') ⬝ (kronecker_map_linear f B B') :=
+begin
+  ext ⟨i, i'⟩ ⟨j, j'⟩,
+  simp only [kronecker_map_linear_apply_apply, mul_apply, ← finset.univ_product_univ,
+    finset.sum_product, kronecker_map],
+  simp_rw [f.map_sum, linear_map.sum_apply, linear_map.map_sum, h_comm],
+end
+
+end kronecker_map
+
+/-! ### Specialization to `matrix.kronecker_map (*)` -/
+
+section kronecker
+
+variables (R)
+
+open_locale matrix
+
+/-- The Kronecker product. This is just a shorthand for `kronecker_map (*)`. Prefer the notation
+`⊗ₖ` rather than this definition. -/
+@[simp] def kronecker [has_mul α] : matrix l m α → matrix n p α → matrix (l × n) (m × p) α :=
+kronecker_map (*)
+
+localized "infix ` ⊗ₖ `:100 := matrix.kronecker_map (*)" in kronecker
+
+@[simp]
+lemma kronecker_apply [has_mul α] (A : matrix l m α) (B : matrix n p α) (i₁ i₂ j₁ j₂) :
+  (A ⊗ₖ B) (i₁, i₂) (j₁, j₂) = A i₁ j₁ * B i₂ j₂ := rfl
+
+/-- `matrix.kronecker` as a bilinear map. -/
+def kronecker_linear [comm_semiring R] [semiring α] [algebra R α] :
+  matrix l m α →ₗ[R] matrix n p α →ₗ[R] matrix (l × n) (m × p) α :=
+kronecker_map_linear (algebra.lmul R α).to_linear_map
+
+/-! What follows is a copy, in order, of every `matrix.kronecker_map` lemma above that has
+hypothese which can be filled by properties of `*`. -/
+
+lemma zero_kronecker [mul_zero_class α] (B : matrix n p α) : (0 : matrix l m α) ⊗ₖ B = 0 :=
+kronecker_map_zero_left _ zero_mul B
+
+lemma kronecker_zero [mul_zero_class α] (A : matrix l m α) : A ⊗ₖ (0 : matrix n p α) = 0 :=
+kronecker_map_zero_right _ mul_zero A
+
+lemma add_kronecker [distrib α] (A₁ A₂ : matrix l m α) (B : matrix n p α) :
+  (A₁ + A₂) ⊗ₖ B = A₁ ⊗ₖ B + A₂ ⊗ₖ B :=
+kronecker_map_add_left _ add_mul _ _ _
+
+lemma kronecker_add [distrib α] (A : matrix l m α) (B₁ B₂ : matrix n p α) :
+  A ⊗ₖ (B₁ + B₂) = A ⊗ₖ B₁ + A ⊗ₖ B₂ :=
+kronecker_map_add_right _ mul_add _ _ _
+
+lemma smul_kronecker [monoid R] [monoid α] [mul_action R α] [is_scalar_tower R α α]
```

Comment:
Should this be a simp-lemma? Same for the next lemma.

Reply:
I think probably not, since we don't make the corresponding lemma for `mul` instead of `kronecker` simp.